reinforcement_learning/rlclientlib/ranking_event.cc

```diff
@@ -56,79 +55,133 @@ namespace reinforcement_learning {
     , _body(body)
   { }
 
+  ranking_event::ranking_event(u::data_buffer& oss, const char* event_id, const char* context,
+    const ranking_response& response, float pass_prob)
+    : event(event_id, pass_prob)
+  {
+    _context = std::string(context);
+    for (auto const &r : response) {
+      _a_vector.push_back(r.action_id);
+      _p_vector.push_back(r.probability);
+    }
+    _model_id = std::string(response.get_model_id());
+  }
+
   ranking_event::ranking_event(ranking_event&& other)
     : event(std::move(other))
     , _body(std::move(other._body))
-  {}
+    , _context(std::move(other._context))
+    , _a_vector(std::move(other._a_vector))
+    , _p_vector(std::move(other._p_vector))
+    , _model_id(std::move(other._model_id))
+  { }
 
   ranking_event& ranking_event::operator=(ranking_event&& other) {
     if (&other != this) {
       event::operator=(std::move(other));
       _body = std::move(other._body);
+      _context = std::move(other._context);
```

Comment:
why do we need another copies of message content here and postpone flatbuffer message building?

Reply:
This is single event data. Flatbuffer is defining an flatbuffer object contains "an array of object". It can only be built when the array is available. There is no way to build it here.

---
In reply to: [228608357](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1653#discussion_r228608357) [](ancestors = 228608357)

Reply:
I mean building of RankingEvent/OutcomeEvent (basically serialize_eventhub_message logic). Why cannot we do in constructor and make reinforcement_learning::ranking event be the holder of RankingEvent only (and the same for outcome_event)?

Reply:
It cannot. We won't serialize RankingEvent/OutComeEvent. What we serialize is RankingEventBatch and OutComeEventBatch. Serialize  RankingEventBatch and "serialize RankingEvent and batch them" are not equivalent. The batch size is determined by event queue size. Batch size in the class is only a max queue size.

---
In reply to: [228634268](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1653#discussion_r228634268) [](ancestors = 228634268)